### PR TITLE
never symlink executables

### PR DESCRIPTION
**Pull request by kalefranz**
A low-level response to https://github.com/conda/conda/issues/3974 and https://github.com/ContinuumIO/anaconda-issues/issues/1392.

This is targeted for 4.4.0 because I'd prefer a change like this go through a full feature canary cycle.

CC @mingwandroid

### CODE REVIEW

**Comment by kalefranz** [comment]
Tests pass here.  Keeping the failing appveyor test because it recorded a 502 error from Cloudflare that's relevant elsewhere.

**Comment by mingwandroid** [comment]
Thanks @kalefranz.

> This is targeted for 4.4.0 because I'd prefer a change like this go through a full feature canary cycle.

I cannot stress enough how important this fix is to both Enterprise (and anyone evaluating Anaconda for that purpose) and also CI systems (I've seen conda-forge workaround this bug in a way that created broken packages, I think/hope they're setting use_symlinks to false now instead). If 4.3 is to support multiuser on Linux then this fix is essential so my opinion is that we need to fast track this.

Could we make a test release of Miniconda and ask as many people as possible (including conda-forge and our QA team) to take it for a prolonged test-drive? I think that would be the best plan of action.

**Comment by mingwandroid** [comment]
One thing I noticed when reviewing, is_executable is determined by checking the executable bit for macOS and Linux. I would like to change this to inspect the binary instead. It is very easy to re-use a small bits of the code I wrote to fix virtualenv for that purpose.

To enable this copy-only code-path we want to ensure the file is a dynamic executable.

**Comment by mingwandroid** [comment]
Either way, before we go ahead with this, I need to check if this copy-only stuff is also necessary for dynamic libraries. From what I've read, it should be, from what I've seen, it isn't but that's not a good enough level of certainty.

**Comment by kalefranz** [comment]
Sure. Just let me know when you're ready to move forward...

**Comment by mingwandroid** [comment]
@kalefranz, I have checked this in detail now.

We only need to do this copy for executables which have an `RPATH` containing `$ORIGIN` on Linux. Currently this PR checks that the executable bit is set will catch everything we need it to catch as well as shell scripts too and things mistakenly set as executable.

Do you want me to port my ELF interpreting code from my `virtualenv` patch and use it to determine precisely if a file needs to be copied or not? It won't be a big speed hit I don't expect, and the saving from not copying would more than make up for it in any case (and of course it would not even get run when hardlinks can be used).

Regardless, please add this to the commit message for future reference (if for no one elses then for my own):

### Start commit message

There has been a lot of confusion regarding whether symlinks are followed for
executables and also for DSOs (dynamic shared objects) and also some confusion
about what `$ORIGIN` means. This is an attempt to clear that up. Everything was
determined from reading the source code for and debugging in `glibc 2.5`.

`-Wl,-z,origin` is irrelevant, it sets `DF_1_ORIGIN` or `DF_ORIGIN` which are
documented to mean "$ORIGIN must be handled."/"Object may use DF_ORIGIN"
respectively. Thing is, neither of those defines can be found referenced
anywhere, just one occurrence each in `glibc/elf/elf.h`. It may be that it is
important in later versions of `glibc` though.

`$ORIGIN` for the executable is resolved as `readlink("/self/proc/exe")`; It
is evaluated once per process and cached in the executable's `link_map->l_origin`.
AFAICT, this is the only time that the dynamic loader ever calls `readlink()`.

This can be seen where `expand_dynamic_string_token()` calls the `DL_DST_REQUIRED
macro. `expand_dynamic_string_token()` is called once for the exe and once for
each DSO. l_origin is a `char*` at offset `0x340` in `link_map`.
`DL_DST_REQUIRED` contains:

```
if ((l)->l_origin == NULL)					      \
  {								      \
    assert ((l)->l_name[0] == '\0');				      \
    (l)->l_origin = _dl_get_origin ();				      \
```

` _dl_get_origin()` begins:
```
const char *
_dl_get_origin (void)
{
  char linkval[PATH_MAX];
  char *result;
  int len;
  INTERNAL_SYSCALL_DECL (err);

len = INTERNAL_SYSCALL (readlink, err, 3, "/proc/self/exe", linkval,
                        sizeof (linkval));
```

Subsequent DSOs set their `l_origin` to dirname(realname) in `_dl_new_object()`.
Crucially they do *not* call `readlink()` because by the time the
`if ((l)->l_origin == NULL)` check happens, `_dl_new_object()` has already set
it.

#### End of commit message.

**Comment by kalefranz** [comment]
repointed to 4.3.x

File: conda/gateways/disk/create.py [hunk 1]
```diff
@@ -178,6 +178,28 @@ def create_hard_link_or_copy(src, dst):
         shutil.copy2(src, dst)
 
 
+def _do_softlink(src, dst):
+    if is_executable(src):
+        # for extra details, see https://github.com/conda/conda/pull/4625#issuecomment-280696371
+        # We only need to do this copy for executables which have an RPATH containing $ORIGIN
+        #   on Linux, so `is_executable()` is currently overly aggressive.
+        # A future optimization will be to copy code from @mingwandroid's virtualenv patch.
```
Comment (kalefranz) [inline conda/gateways/disk/create.py]: @mingwandroid Are you fine with the above for now?

Comment (mingwandroid) [inline conda/gateways/disk/create.py]: Not really, I'm a big fan of detailed commit messages, and would rather you added my commentary to the commit message. Think of the developers on air-planes (ones without internet).

Comment (mingwandroid) [inline conda/gateways/disk/create.py]: `git commit --amend ; git push --force` ftw ..

Comment (kalefranz) [inline conda/gateways/disk/create.py]: Oh I did that too already.

https://github.com/conda/conda/pull/4625/commits/b253c467bed429969e2ebd9e7d26a8f0cbe01d25

The full message is there.

I just meant, are you ok with being overly aggressive right now, and then adding in more optimized logic later?

Comment (mingwandroid) [inline conda/gateways/disk/create.py]: Ah ok, yes completely fine with that.

File: conda/gateways/disk/create.py [hunk 2]
```diff
@@ -178,6 +178,30 @@ def create_hard_link_or_copy(src, dst):
         shutil.copy2(src, dst)
 
 
+def _do_softlink(src, dst):
+    if islink(src):
+        symlink(src, dst)
```
Comment (kalefranz) [inline conda/gateways/disk/create.py]: @mingwandroid I added the two above lines.  Are you fine with this still?  If not, we might have to tighten down `is_executable()` a lot more.

File: conda/gateways/disk/permissions.py [hunk 3]
```diff
@@ -64,3 +65,9 @@ def make_executable(path):
         chmod(path, S_IMODE(mode) | S_IXUSR | S_IXGRP | S_IXOTH)
     else:
         log.error("Cannot make path '%s' executable", path)
+
+
+def is_executable(path):
+    if isfile(path):  # for now, leave out `and not islink(path)`
+        return path.endswith(('.exe', '.bat')) if on_win else access(path, X_OK)
+    return False
```
Comment (kalefranz) [inline conda/gateways/disk/permissions.py]: @mingwandroid Also slight modification to this function.  It interacts with `_do_softlink()`.  One thing I've learned for sure being bitten too many times by it... `os.access()` just doesn't work on Windows.

Comment (kalefranz) [inline conda/gateways/disk/permissions.py]: Just linux?  OSX doesn't need this?

Comment (kalefranz) [inline conda/gateways/disk/permissions.py]: Made one more push with a special function right next to `_do_push`.

Comment (mingwandroid) [inline conda/gateways/disk/permissions.py]: AFAIK, macOS doesn't need this, but it may be safer this way.

Comment (kalefranz) [inline conda/gateways/disk/permissions.py]: Safety is good. I like safe :)

**Pull request merged**